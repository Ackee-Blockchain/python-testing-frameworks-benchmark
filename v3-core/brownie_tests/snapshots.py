import json
from pathlib import Path


def match_snapshot(value, file_path, ident):
    filename = Path(__file__).parent / "__snapshots__" / f"{Path(file_path).name}.snap"
    with open(filename, "r") as f:
        expected = json.load(f)
        assert value == int(expected[ident])
