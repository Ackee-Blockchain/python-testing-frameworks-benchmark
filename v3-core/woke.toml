[detectors]
exclude = []
ignore_paths = [ "node_modules", ".woke-build", "venv", "lib",]

[testing]
cmd = "hardhat"

[testing.anvil]
cmd_args = "--prune-history 100 --steps-tracing --silent"

[testing.ganache]
cmd_args = "-k istanbul -q"

[testing.hardhat]
cmd_args = ""

[compiler.solc]
ignore_paths = [ "node_modules", ".woke-build", "venv", "lib", "audits",]
include_paths = [ "node_modules",]

[compiler.solc.optimizer]
enabled = true
runs = 200
