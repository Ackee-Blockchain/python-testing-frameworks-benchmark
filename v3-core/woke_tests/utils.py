import dataclasses
import decimal
import math
from decimal import Decimal
from enum import IntEnum

from eth_abi.packed import encode_packed
from eth_utils import keccak
from pytypes.contracts.test.MockTimeUniswapV3Pool import MockTimeUniswapV3Pool
from pytypes.contracts.test.MockTimeUniswapV3PoolDeployer import \
    MockTimeUniswapV3PoolDeployer
from pytypes.contracts.test.TestERC20 import TestERC20
from pytypes.contracts.test.TestUniswapV3Callee import TestUniswapV3Callee
from pytypes.contracts.UniswapV3Factory import UniswapV3Factory
from woke.testing import *

decimal.setcontext(decimal.Context(prec=40))

MAX_UINT_256 = 2**256 - 1
MAX_UINT_128 = 2**128 - 1
MIN_SQRT_RATIO = 4295128739
MAX_SQRT_RATIO = 1461446703485210103287273052203988822378723970342
TEST_POOL_START_TIME = 1601906400


class FeeAmount(IntEnum):
    LOW = 500
    MEDIUM = 3000
    HIGH = 10000


class TickSpacings(IntEnum):
    LOW = 10
    MEDIUM = 60
    HIGH = 200


def get_min_tick(tick_spacing):
    return math.ceil(-887272 / tick_spacing) * tick_spacing


def get_max_tick(tick_spacing):
    return math.floor(887272 / tick_spacing) * tick_spacing


def get_max_liquidity_per_tick(tick_spacing):
    return Decimal(MAX_UINT_128) / int(
        (get_max_tick(tick_spacing) - get_min_tick(tick_spacing)) / tick_spacing + 1
    )


def expand_to_18_decimals(n):
    return n * 10**18


def encode_price_sqrt(reserve1, reserve0):
    return int(
        (Decimal.from_float(reserve1) / Decimal.from_float(reserve0)).sqrt() * 2**96
    )


def get_position_key(address: str, lower_tick: int, upper_tick: int) -> bytes:
    packed_data = encode_packed(
        ["address", "int24", "int24"], [address, lower_tick, upper_tick]
    )
    return keccak(packed_data)


def check_observation_equals(observation, expected_observation):
    assert dataclasses.asdict(observation) == dataclasses.asdict(expected_observation)


def check_tick_is_clear(pool, tick):
    tick_info = pool.ticks(tick)
    assert tick_info.liquidityGross == 0
    assert tick_info.feeGrowthOutside0X128 == 0
    assert tick_info.feeGrowthOutside1X128 == 0
    assert tick_info.liquidityNet == 0


def check_tick_is_not_clear(pool, tick):
    assert pool.ticks(tick).liquidityGross != 0


def create_pool(fee, tick_spacing, first_token, second_token, factory):
    mock_time_pool_deployer = MockTimeUniswapV3PoolDeployer.deploy(
        from_=default_chain.accounts[0]
    )
    tx = mock_time_pool_deployer.deploy_(
        factory.address,
        first_token.address,
        second_token.address,
        fee.value,
        tick_spacing.value,
        from_=default_chain.accounts[0],
    )

    return MockTimeUniswapV3Pool(tx.events[0].pool)


class PoolHelper:
    def __init__(
        self,
        token0: TestERC20,
        token1: TestERC20,
        factory: UniswapV3Factory,
        pool: MockTimeUniswapV3Pool,
        tick_spacing,
        swap_target: TestUniswapV3Callee,
    ):
        self.token0 = token0
        self.token1 = token1
        self.factory = factory
        self.swap_target = swap_target
        self.pool = pool
        self.tick_spacing = tick_spacing
        self.min_tick = get_min_tick(tick_spacing)
        self.max_tick = get_max_tick(tick_spacing)

    def swap(self, input_token, amount_in, amount_out, to, sqrt_price_limit_x96=None):
        exact_input = amount_out == 0
        method = None
        if input_token == self.token0:
            method = (
                self.swap_target.swapExact0For1
                if exact_input
                else self.swap_target.swap0ForExact1
            )
        else:
            method = (
                self.swap_target.swapExact1For0
                if exact_input
                else self.swap_target.swap1ForExact0
            )

        if sqrt_price_limit_x96 is None:
            sqrt_price_limit_x96 = (
                MIN_SQRT_RATIO + 1 if input_token == self.token0 else MAX_SQRT_RATIO - 1
            )
        input_token.approve(self.swap_target.address, MAX_UINT_256)
        to_address = to if isinstance(to, str) else to.address
        return method(
            self.pool.address,
            amount_in if exact_input else amount_out,
            to_address,
            sqrt_price_limit_x96,
        )

    def mint(self, recipient, tick_lower, tick_upper, liquidity):
        self.token0.approve(self.swap_target.address, MAX_UINT_256)
        self.token1.approve(self.swap_target.address, MAX_UINT_256)
        return self.swap_target.mint(
            self.pool.address, recipient, tick_lower, tick_upper, liquidity
        )

    def flash(self, amount0, amount1, to, pay0=None, pay1=None):
        fee = self.pool.fee()
        if pay0 is None:
            pay0 = (amount0 * fee + 10**6 - 1) // 10**6 + amount0
        if pay1 is None:
            pay1 = (amount1 * fee + 10**6 - 1) // 10**6 + amount1
        return self.swap_target.flash(
            self.pool.address,
            to if isinstance(to, str) else to.address,
            amount0,
            amount1,
            pay0,
            pay1,
        )

    def swap_to_sqrt_price(self, input_token, target_price, to):
        method = (
            self.swap_target.swapToLowerSqrtPrice
            if input_token == self.token0
            else self.swap_target.swapToHigherSqrtPrice
        )
        input_token.approve(self.swap_target.address, MAX_UINT_256)
        to_address = to if isinstance(to, str) else to.address
        return method(self.pool.address, target_price, to_address)

    def swap_to_lower_price(self, sqrt_price_x96, to):
        return self.swap_to_sqrt_price(self.token0, sqrt_price_x96, to)

    def swap_to_higher_price(self, sqrt_price_x96, to):
        return self.swap_to_sqrt_price(self.token1, sqrt_price_x96, to)

    def swap_exact_0_for_1(self, amount, to, sqrt_price_limit_x96=None):
        return self.swap(self.token0, amount, 0, to, sqrt_price_limit_x96)

    def swap_0_for_exact_1(self, amount, to, sqrt_price_limit_x96=None):
        return self.swap(self.token0, 0, amount, to, sqrt_price_limit_x96)

    def swap_exact_1_for_0(self, amount, to, sqrt_price_limit_x96=None):
        return self.swap(self.token1, amount, 0, to, sqrt_price_limit_x96)

    def swap_1_for_exact_0(self, amount, to, sqrt_price_limit_x96=None):
        return self.swap(self.token1, 0, amount, to, sqrt_price_limit_x96)
