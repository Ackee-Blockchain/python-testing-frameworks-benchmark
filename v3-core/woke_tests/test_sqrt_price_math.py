import pytest
from pytypes.contracts.test.SqrtPriceMathTest import SqrtPriceMathTest
from woke.testing import *

import woke_tests.utils as utils
from woke_tests.snapshots import match_snapshot
from woke_tests.utils import MAX_UINT_128, MAX_UINT_256


@pytest.fixture(scope="module")
def sqrt_price_math():
    default_chain.set_default_accounts(default_chain.accounts[0])
    sqrt_price_math = SqrtPriceMathTest.deploy(from_=default_chain.accounts[0])

    return sqrt_price_math


@pytest.fixture(scope="module", autouse=True)
def chain():
    with default_chain.connect():
        yield default_chain


class TestGetNextSqrtPriceFromInput:
    def test_zero_price(self, sqrt_price_math):
        """
        fails if price is zero
        """
        with must_revert():
            sqrt_price_math.getNextSqrtPriceFromInput(
                0, 0, int(utils.expand_to_18_decimals(1) / 10), False
            )

    def test_zero_liquidity(self, sqrt_price_math):
        """
        fails if liquidity is zero
        """
        with must_revert():
            sqrt_price_math.getNextSqrtPriceFromInput(
                1, 0, int(utils.expand_to_18_decimals(1) / 10), True
            )

    def test_overflow_price(self, sqrt_price_math):
        """
        fails if input amount overflows the price
        """
        price = 2**160 - 1
        liquidity = 1024
        amount_in = 1024
        with must_revert():
            sqrt_price_math.getNextSqrtPriceFromInput(
                price, liquidity, amount_in, False
            )

    def test_underflow_price(self, sqrt_price_math):
        """
        any input amount cannot underflow the price
        """
        price = 1
        liquidity = 1
        amount_in = 2**255
        assert (
            sqrt_price_math.getNextSqrtPriceFromInput(price, liquidity, amount_in, True)
            == 1
        )

    def test_zero_amount_in_zero_for_one_true(self, sqrt_price_math):
        """
        returns input price if amount in is zero and zeroForOne = true
        """
        price = utils.encode_price_sqrt(1, 1)
        assert (
            sqrt_price_math.getNextSqrtPriceFromInput(
                price, int(utils.expand_to_18_decimals(1) / 10), 0, True
            )
            == price
        )

    def test_zero_amount_in_zero_for_one_false(self, sqrt_price_math):
        """
        returns input price if amount in is zero and zeroForOne = false
        """
        price = utils.encode_price_sqrt(1, 1)
        assert (
            sqrt_price_math.getNextSqrtPriceFromInput(
                price, int(utils.expand_to_18_decimals(1) / 10), 0, False
            )
            == price
        )

    def test_input_max_inputs(self, sqrt_price_math):
        """
        returns the minimum price for max inputs
        """
        sqrt_p = 2**160 - 1
        liquidity = MAX_UINT_128
        # had to sub 1 because of rounding vs ethers BigNumber
        max_amount_no_overflow = MAX_UINT_256 - int((liquidity << 96) / sqrt_p - 1)
        assert (
            sqrt_price_math.getNextSqrtPriceFromInput(
                sqrt_p, liquidity, max_amount_no_overflow, True
            )
            == 1
        )

    def test_small_input_amount_token1(self, sqrt_price_math):
        """
        input amount of 0.1 token1
        """
        assert (
            sqrt_price_math.getNextSqrtPriceFromInput(
                utils.encode_price_sqrt(1, 1),
                utils.expand_to_18_decimals(1),
                int(utils.expand_to_18_decimals(1) / 10),
                False,
            )
            == 87150978765690771352898345369
        )

    def test_small_input_amount_token0(self, sqrt_price_math):
        """
        input amount of 0.1 token0
        """
        assert (
            sqrt_price_math.getNextSqrtPriceFromInput(
                utils.encode_price_sqrt(1, 1),
                utils.expand_to_18_decimals(1),
                int(utils.expand_to_18_decimals(1) / 10),
                True,
            )
            == 72025602285694852357767227579
        )

    def test_large_liquidity_ten(self, sqrt_price_math):
        """
        amountIn > type(uint96).max and zeroForOne = true
        """
        assert (
            sqrt_price_math.getNextSqrtPriceFromInput(
                utils.encode_price_sqrt(1, 1),
                utils.expand_to_18_decimals(10),
                2**100,
                True,
            )
            == 624999999995069620
        )

    def test_half_max_uint_256(self, sqrt_price_math):
        """
        can return 1 with enough amountIn and zeroForOne = true
        """
        assert (
            sqrt_price_math.getNextSqrtPriceFromInput(
                utils.encode_price_sqrt(1, 1), 1, int(MAX_UINT_256 / 2), True
            )
            == 1
        )

    def test_zero_for_one_true_gas(self, sqrt_price_math):
        """
        zeroForOne = true gas
        """
        match_snapshot(
            sqrt_price_math.getGasCostOfGetNextSqrtPriceFromInput(
                utils.encode_price_sqrt(1, 1),
                utils.expand_to_18_decimals(1),
                int(utils.expand_to_18_decimals(1) / 10),
                True,
            ),
            __file__,
            "getGasCostOfGetNextSqrtPriceFromInput_test_zero_for_one_true_gas",
        )

    def test_zero_for_one_false_gas(self, sqrt_price_math):
        """
        zeroForOne = false gas
        """
        match_snapshot(
            sqrt_price_math.getGasCostOfGetNextSqrtPriceFromInput(
                utils.encode_price_sqrt(1, 1),
                utils.expand_to_18_decimals(1),
                int(utils.expand_to_18_decimals(1) / 10),
                False,
            ),
            __file__,
            "getGasCostOfGetNextSqrtPriceFromInput_test_zero_for_one_false_gas",
        )


class TestGetNextSqrtPriceFromOutput(object):
    def test_zero_price(self, sqrt_price_math):
        """
        fails if price is zero
        """
        with must_revert():
            sqrt_price_math.getNextSqrtPriceFromOutput(
                0, 0, int(utils.expand_to_18_decimals(1) / 10), False
            )

    def test_zero_liquidity(self, sqrt_price_math):
        """
        fails if liquidity is zero
        """
        with must_revert():
            sqrt_price_math.getNextSqrtPriceFromOutput(
                1, 0, int(utils.expand_to_18_decimals(1) / 10), True
            )

    def test_virtual_reserves_token0(self, sqrt_price_math):
        """
        fails if output amount is exactly the virtual reserves of token0
        """
        price = 20282409603651670423947251286016
        liquidity = 1024
        amount_out = 4
        with must_revert():
            sqrt_price_math.getNextSqrtPriceFromOutput(
                price, liquidity, amount_out, False
            )

    def test_greater_than_virtual_reserves_token0(self, sqrt_price_math):
        """
        fails if output amount is greater than virtual reserves of token0
        """
        price = 20282409603651670423947251286016
        liquidity = 1024
        amount_out = 5
        with must_revert():
            sqrt_price_math.getNextSqrtPriceFromOutput(
                price, liquidity, amount_out, False
            )

    def test_greater_than_virtual_reserves_token1(self, sqrt_price_math):
        """
        fails if output amount is greater than virtual reserves of token1
        """
        price = 20282409603651670423947251286016
        liquidity = 1024
        amount_out = 262145
        with must_revert():
            sqrt_price_math.getNextSqrtPriceFromOutput(
                price, liquidity, amount_out, True
            )

    def test_virtual_reserves_token1(self, sqrt_price_math):
        """
        fails if output amount is exactly the virtual reserves of token1
        """
        price = 20282409603651670423947251286016
        liquidity = 1024
        amount_out = 262144
        with must_revert():
            sqrt_price_math.getNextSqrtPriceFromOutput(
                price, liquidity, amount_out, True
            )

    def test_just_less_than_virtual_reserves_token1(self, sqrt_price_math):
        """
        succeeds if output amount is just less than the virtual reserves of token1
        """
        price = 20282409603651670423947251286016
        liquidity = 1024
        amount_out = 262143
        sqrtQ = sqrt_price_math.getNextSqrtPriceFromOutput(
            price, liquidity, amount_out, True
        )
        assert sqrtQ == 77371252455336267181195264

    def test_puzzling_echidna(self, sqrt_price_math):
        """
        puzzling echidna test
        """
        price = 20282409603651670423947251286016
        liquidity = 1024
        amount_out = 4
        with must_revert():
            sqrt_price_math.getNextSqrtPriceFromOutput(
                price, liquidity, amount_out, False
            )

    def test_zero_amount_in_zero_for_one(self, sqrt_price_math):
        """
        returns input price if amount in is zero and zeroForOne = true
        """
        price = utils.encode_price_sqrt(1, 1)
        assert (
            sqrt_price_math.getNextSqrtPriceFromOutput(
                price, int(utils.expand_to_18_decimals(1) / 10), 0, True
            )
            == price
        )

    def test_zero_amount_in_one_for_zero(self, sqrt_price_math):
        """
        returns input price if amount in is zero and zeroForOne = false
        """
        price = utils.encode_price_sqrt(1, 1)
        assert (
            sqrt_price_math.getNextSqrtPriceFromOutput(
                price, int(utils.expand_to_18_decimals(1) / 10), 0, False
            )
            == price
        )

    def test_output_amount_of_01_token1(self, sqrt_price_math):
        """
        output amount of 0.1 token1
        """
        sqrtQ = sqrt_price_math.getNextSqrtPriceFromOutput(
            utils.encode_price_sqrt(1, 1),
            utils.expand_to_18_decimals(1),
            int(utils.expand_to_18_decimals(1) / 10),
            False,
        )
        assert sqrtQ == 88031291682515930659493278152

    def test_output_amount_of_01_token2(self, sqrt_price_math):
        """
        output amount of 0.1 token1
        """
        sqrtQ = sqrt_price_math.getNextSqrtPriceFromOutput(
            utils.encode_price_sqrt(1, 1),
            utils.expand_to_18_decimals(1),
            int(utils.expand_to_18_decimals(1) / 10),
            True,
        )
        assert sqrtQ == 71305346262837903834189555302

    def test_impossible_amount_out_zero_for_one(self, sqrt_price_math):
        """
        reverts if amountOut is impossible in zero for one direction
        """
        with must_revert():
            sqrt_price_math.getNextSqrtPriceFromOutput(
                utils.encode_price_sqrt(1, 1), 1, MAX_UINT_256, True
            )

    def test_impossible_amount_out_one_for_zero(self, sqrt_price_math):
        """
        reverts if amountOut is impossible in one for zero direction
        """
        with must_revert():
            sqrt_price_math.getNextSqrtPriceFromOutput(
                utils.encode_price_sqrt(1, 1), 1, MAX_UINT_256, False
            )

    def test_zero_for_one_true_gas(self, sqrt_price_math):
        """
        gas cost for getNextSqrtPriceFromOutput where zeroForOne = true
        """
        match_snapshot(
            sqrt_price_math.getGasCostOfGetNextSqrtPriceFromOutput(
                utils.encode_price_sqrt(1, 1),
                utils.expand_to_18_decimals(1),
                int(utils.expand_to_18_decimals(1) / 10),
                True,
            ),
            __file__,
            "getGasCostOfGetNextSqrtPriceFromOutput_test_zero_for_one_true_gas",
        )

    def test_zero_for_one_false_gas(self, sqrt_price_math):
        """
        gas cost for getNextSqrtPriceFromOutput where zeroForOne = false
        """
        match_snapshot(
            sqrt_price_math.getGasCostOfGetNextSqrtPriceFromOutput(
                utils.encode_price_sqrt(1, 1),
                utils.expand_to_18_decimals(1),
                int(utils.expand_to_18_decimals(1) / 10),
                False,
            ),
            __file__,
            "getGasCostOfGetNextSqrtPriceFromOutput_test_zero_for_one_false_gas",
        )


class TestGetAmount0Delta(object):
    def test_zero_liquidity(self, sqrt_price_math):
        """
        returns 0 if liquidity is 0
        """
        amount0 = sqrt_price_math.getAmount0Delta(
            utils.encode_price_sqrt(1, 1), utils.encode_price_sqrt(2, 1), 0, True
        )
        assert amount0 == 0

    def test_equal_prices(self, sqrt_price_math):
        """
        returns 0 if prices are equal
        """
        amount0 = sqrt_price_math.getAmount0Delta(
            utils.encode_price_sqrt(1, 1), utils.encode_price_sqrt(1, 1), 0, True
        )
        assert amount0 == 0

    def test_price_1_to_1_21(self, sqrt_price_math):
        """
        returns 0.1 amount1 for price of 1 to 1.21
        """
        amount0 = sqrt_price_math.getAmount0Delta(
            utils.encode_price_sqrt(1, 1),
            utils.encode_price_sqrt(121, 100),
            utils.expand_to_18_decimals(1),
            True,
        )
        assert amount0 == 90909090909090910

        amount0_rounded_down = sqrt_price_math.getAmount0Delta(
            utils.encode_price_sqrt(1, 1),
            utils.encode_price_sqrt(121, 100),
            utils.expand_to_18_decimals(1),
            False,
        )
        assert amount0_rounded_down == amount0 - 1

    def test_overflow_prices(self, sqrt_price_math):
        """
        works for prices that overflow
        """
        amount0_up = sqrt_price_math.getAmount0Delta(
            utils.encode_price_sqrt(2**90, 1),
            utils.encode_price_sqrt(2**96, 1),
            utils.expand_to_18_decimals(1),
            True,
        )
        amount0_down = sqrt_price_math.getAmount0Delta(
            utils.encode_price_sqrt(2**90, 1),
            utils.encode_price_sqrt(2**96, 1),
            utils.expand_to_18_decimals(1),
            False,
        )
        assert amount0_up == amount0_down + 1

    def test_cost_true_gas(self, sqrt_price_math):
        """
        gas cost for amount0 where roundUp = true
        """
        match_snapshot(
            sqrt_price_math.getGasCostOfGetAmount0Delta(
                utils.encode_price_sqrt(100, 121),
                utils.encode_price_sqrt(1, 1),
                utils.expand_to_18_decimals(1),
                True,
            ),
            __file__,
            "getGasCostOfGetAmount0Delta_test_gas_cost_true",
        )

    def test_cost_false_gas(self, sqrt_price_math):
        """
        gas cost for amount0 where roundUp = false
        """
        match_snapshot(
            sqrt_price_math.getGasCostOfGetAmount0Delta(
                utils.encode_price_sqrt(100, 121),
                utils.encode_price_sqrt(1, 1),
                utils.expand_to_18_decimals(1),
                False,
            ),
            __file__,
            "getGasCostOfGetAmount0Delta_test_gas_cost_false",
        )


class TestGetAmount1Delta(object):
    def test_zero_liquidity(self, sqrt_price_math):
        """
        returns 0 if liquidity is 0
        """
        amount1 = sqrt_price_math.getAmount1Delta(
            utils.encode_price_sqrt(1, 1), utils.encode_price_sqrt(2, 1), 0, True
        )
        assert amount1 == 0

    def test_equal_prices(self, sqrt_price_math):
        """
        returns 0 if prices are equal
        """
        amount1 = sqrt_price_math.getAmount0Delta(
            utils.encode_price_sqrt(1, 1), utils.encode_price_sqrt(1, 1), 0, True
        )
        assert amount1 == 0

    def test_price_1_to_1_21(self, sqrt_price_math):
        """
        returns 0.1 amount1 for price of 1 to 1.21
        """
        amount1 = sqrt_price_math.getAmount1Delta(
            utils.encode_price_sqrt(1, 1),
            utils.encode_price_sqrt(121, 100),
            utils.expand_to_18_decimals(1),
            True,
        )
        assert amount1 == 100000000000000000
        amount1_rounded_down = sqrt_price_math.getAmount1Delta(
            utils.encode_price_sqrt(1, 1),
            utils.encode_price_sqrt(121, 100),
            utils.expand_to_18_decimals(1),
            False,
        )
        assert amount1_rounded_down == amount1 - 1

    def test_round_up_true_gas(self, sqrt_price_math):
        """
        gas cost for amount0 where roundUp = true
        """
        match_snapshot(
            sqrt_price_math.getGasCostOfGetAmount0Delta(
                utils.encode_price_sqrt(100, 121),
                utils.encode_price_sqrt(1, 1),
                utils.expand_to_18_decimals(1),
                True,
            ),
            __file__,
            "getGasCostOfGetAmount0Delta_test_round_up_true",
        )

    def test_round_up_false_gas(self, sqrt_price_math):
        """
        gas cost for amount0 where roundUp = false
        """
        match_snapshot(
            sqrt_price_math.getGasCostOfGetAmount0Delta(
                utils.encode_price_sqrt(100, 121),
                utils.encode_price_sqrt(1, 1),
                utils.expand_to_18_decimals(1),
                False,
            ),
            __file__,
            "getGasCostOfGetAmount0Delta_test_round_up_false",
        )


class TestSwapComputation(object):
    def test_sqrtp_sqrtq_overflows(self, sqrt_price_math):
        """
        sqrtP * sqrtQ overflows
        """
        sqrtP = 1025574284609383690408304870162715216695788925244
        liquidity = 50015962439936049619261659728067971248
        zeroForOne = True
        amountIn = 406

        sqrtQ = sqrt_price_math.getNextSqrtPriceFromInput(
            sqrtP, liquidity, amountIn, zeroForOne
        )
        assert sqrtQ == 1025574284609383582644711336373707553698163132913

        amount0Delta = sqrt_price_math.getAmount0Delta(sqrtQ, sqrtP, liquidity, True)
        assert amount0Delta == 406
