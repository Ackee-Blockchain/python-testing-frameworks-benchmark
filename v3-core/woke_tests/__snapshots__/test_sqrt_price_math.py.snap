{
    "getGasCostOfGetAmount0Delta_test_gas_cost_true":"628",
    "getGasCostOfGetAmount0Delta_test_gas_cost_false":"496",
    "getGasCostOfGetAmount0Delta_test_round_up_true":"628",
    "getGasCostOfGetAmount0Delta_test_round_up_false":"496",
    "getGasCostOfGetNextSqrtPriceFromOutput_test_zero_for_one_true_gas":"441",
    "getGasCostOfGetNextSqrtPriceFromOutput_test_zero_for_one_false_gas":"866",
    "getGasCostOfGetNextSqrtPriceFromInput_test_zero_for_one_true_gas":"771",
    "getGasCostOfGetNextSqrtPriceFromInput_test_zero_for_one_false_gas":"536"
}